DEPS

```diff
@@ -18,6 +22,7 @@ deps = {
   "third_party/ethash/src": "https://github.com/chfast/ethash.git@e4a15c3d76dc09392c7efd3e30d84ee3b871e9ce",
   "third_party/bitcoin-core/src": "https://github.com/bitcoin/bitcoin.git@95ea54ba089610019a74c1176a2c7c0dba144b1c",
   "third_party/argon2/src": "https://github.com/P-H-C/phc-winner-argon2.git@62358ba2123abd17fccf2a108a301d4b52c01a7c",
+  "third_party/rust/cxx": "https://github.com/dtolnay/cxx.git@refs/tags/" + Var('brave_cxx_version'),
```

Comment:
That can be changed by pushing to the repo so we should have a commit hash here instead.

Reply:
Commit hash is problematic because the crate and the repo have to stay in sync and we can't use sha for cargo install afaik

Reply:
Also specifying a hit repo for cargo isn't an option because that will cause the same build issue on windows that we had building it ourselves from the hit repo

Reply:
For cargo, are we just specifying the version in the `Cargo.toml` file?

Using a hash here would be problematic because we would have to update both `DEPS` and then `Cargo.toml` at the same time, and that could lead to them getting out of sync?

Reply:
It's not specified in a Cargo.toml file, we call `cargo install` directly with the version and yes it could lead to them getting out of sync. This way we specify the version for both in one place https://github.com/brave/brave-core/pull/10847/files/12586d5a1453d0a17af46276d9efca6ea1763f51#diff-4cc955bcf0b472353786cc8d8b7d346fb432c091bed099de884885e938392814R4

Reply:
We can definitely `cargo install` from a commit hash:

```
> cargo install --help
Install a Rust binary. Default location is $HOME/.cargo/bin

USAGE:
    cargo install [OPTIONS] [--] [crate]...
    
OPTIONS:
    ...
    --git <URL>                 Git URL to install the specified crate from
    --branch <BRANCH>           Branch to use when installing from git
    --tag <TAG>                 Tag to use when installing from git
    --rev <SHA>                 Specific commit to use when installing from git
    ...
```

@bridiver what was the Windows build issue?

Reply:
@antonok-edm it would throw an error about symlink gitconfig option not supported:
```
error: failed to run custom build command for `cxxbridge-cmd v1.0.56 (C:\PRJ\brave-browser\src\brave\third_party\rust\cxx\gen\cmd)`

Caused by:
  process didn't exit successfully: `C:\PRJ\brave-browser\src\out\Component\cxxbridge\release\build\cxxbridge-cmd-9323a71da3abb466\build-script-build` (exit code: 1)
  --- stderr

  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  When building `cxx` from a git clone, git's symlink support needs
  to be enabled on platforms that have it off by default (Windows).
  Either use:

     $ git config --global core.symlinks true

  prior to cloning, or else use:

     $ git clone -c core.symlinks=true https://github.com/dtolnay/cxx

  for the clone.

  Symlinks are only required when compiling locally from a clone of
  the git repository---they are NOT required when building `cxx` as
  a Cargo-managed (possibly transitive) build dependency downloaded
  through crates.io.
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: build failed, waiting for other jobs to finish...
error: build failed
```

Reply:
@bsclifton weird! And using either of the suggested `git` commands is not an option either?

Reply:
@antonok-edm also that doesn't solve the problem because it can't be a git checkout.

Reply:
if we could use a git checkout we wouldn't have to cargo install in the first place

Reply:
@fmarier @antonok-edm is there something similar to Cargo.lock here we can use to make sure it doesn't get changed? I noticed it uses `--locked` to install these, but it's not clear to me where/how that Cargo.lock is maintained. Maybe we can commit it?

Reply:
that doesn't solve the git tag issue though

Reply:
would the following work?
1. change this to a commit hash
2. cargo install using `--path`
3. reference cxx in our `Cargo.toml` via path

Reply:
Is there a reason we can't add the binary to the rust_deps blob like we do for `cbindgen` and `cargo audit`?

I guess I also don't understand why you need to install from git instead of from crates.io, which should at least have immutable version numbers?

Reply:
my understanding based on the changes to `scripts/download_rust_deps.py` that it would be included in the rust deps blob?

there are three separate uses of the cxx repo:
1. cargo install of `cxxbridge-cmd`
2. having `cxx.cc` and `cxx.h` available for one of our C++ compilation targets
3. using `cxx` as a cargo dependency

we have the following constraints:
a. we want all three uses to use the same revision of the cxx repo
b. have it reliably be pinned to a specific content hash ( as requested at the start of this thread )

additionally, in the ideal case we would have said content hash specified in a single place.

due to 2 there isn't really a good way around us checking out the cxx git repo directly at which point I think using path specifiers for our cargo dep and installs as I mentioned above is easiest

Reply:
Thanks for explaining in more detail!

I'm still confused by step two. The crates.io package definitely includes the c++ utility source
```
$ curl -sL https://crates.io/api/v1/crates/cxx/1.0.56/download | tar tvz | grep 'cxx.[hc]'
-rw-r--r-- 0/0             606 1973-11-29 13:33 cxx-1.0.56/book/css/cxx.css
-rw-r--r-- 0/0           27930 1973-11-29 13:33 cxx-1.0.56/include/cxx.h
-rw-r--r-- 0/0           30380 1973-11-29 13:33 cxx-1.0.56/src/cxx.cc
```
Is is issue finding the package source reliably when that's kind of a cargo internal? Not trusting crates.io as a source? Could we use `cargo install --locked` for 1 and add a custom build step verify an in-tree hash of `cxx.h` and `cxx.cc`?

Reply:
i can't speak authoritatively as I didn't implement the current state - but I suspect it's more the former. finding those files via internal cargo paths seems much more error prone / hacky IMO

probably? i'm having trouble seeing the advantage though, checking out the repo + installing / referencing via path seems dead simple

Reply:
> Is there a reason we can't add the binary to the rust_deps blob like we do for `cbindgen` and `cargo audit`?

That is exactly what we do
 
> I guess I also don't understand why you need to install from git instead of from crates.io, which should at least have immutable version numbers?

we specifically can't install from git which is why we're using a version instead of a hash. If you try to build on windows from a git checkout it fails because windows doesn't have symlink support turned on by default

Reply:
> i can't speak authoritatively as I didn't implement the current state - but I suspect it's more the former. finding those files via internal cargo paths seems much more error prone / hacky IMO
The .cc file doesn't exist in the cargo install, only the .h

> probably? i'm having trouble seeing the advantage though, checking out the repo + installing / referencing via path seems dead simple
We can't install from the repo as mentioned above. That's what we did originally and it fails on windows because of the symlink thing which is why we did this instead

Reply:
Attempting to summarise out-of-band discussion:

- We need the `cxxbridge` command available because we want gn, not cargo, to drive building the c++ side of the code. Likewise we need `cxx.h` and `cxx.cc` from the cxx crate to provide the implementation of the rust type wrappers.
- We cannot use `cargo install` to build `cxxbridge` from the `cxx` crate or its source repo because of the symlink issue the crate has on Windows.
- We'd like `rust_deps` to include the necessary stuff, but for changes in the required version elsewhere in the repo to override/update that to simplify development.

A suggested solution is to use the `cxx` crate as the source of `cxx.cc` and `cxx.h`. We can set the required version in e.g. `build/rust/Cargo.toml`/`.lock` and have `download_rust_deps.py` run `cargo fetch` to ensure that source code is available in the repository cache before the build starts. The same script can then parse out the `cxx` crate version and `cargo install` the corresponding version of the `cxxbridge-cmd` crate from the crates.io package, which doesn't have the symlink issue building on windows.

Reply:
@fmarier @rillian should be fixed with https://github.com/brave/brave-core/pull/10847/commits/64143d3f43d697a741e60ba51fa3dd38b4e03372 in case you wanted to re-review 😄